src/node_manager/mod.rs

```diff
@@ -111,19 +109,31 @@ impl NodeManager {
             self.nodes.clone()
         };
 
-        for mut node in nodes {
-            node.url.set_path(path);
-            node.url.set_query(query);
-            nodes_with_modified_url.insert(node);
+        // Add remaining nodes in random order
+        for node in nodes_random_order {
+            if !nodes_with_modified_url.iter().any(|n| n.url == node.url) {
+                nodes_with_modified_url.push(node);
+            }
         }
 
         // remove disabled nodes
         nodes_with_modified_url.retain(|n| !n.disabled);
+
         if nodes_with_modified_url.is_empty() {
             return Err(crate::Error::SyncedNodePoolEmpty);
         }
 
-        Ok(nodes_with_modified_url.into_iter().collect())
+        // Set path and query parameters
+        nodes_with_modified_url = nodes_with_modified_url
+            .into_iter()
+            .map(|mut node| {
+                node.url.set_path(path);
+                node.url.set_query(query);
+                node
+            })
+            .collect();
```

Comment:
Use `.iter_mut().for_each()` instead of `.into_iter().collect()`, less allocations